### PR TITLE
Command debug

### PR DESCRIPTION
**Pull request by BrownGenius**
Hi team, since we are experimenting with more complex autonomous/assist routines, I added some debug/trace code so that we can see what is happening in the sequential commands.  E.g. this is the output to the console when executing the AutoScore routine:

```
bhs.devilbotz.commands.assist.AutoScore:Start
bhs.devilbotz.commands.arm.ArmToPosition:initialize
bhs.devilbotz.commands.arm.ArmToPosition:end
bhs.devilbotz.commands.gripper.GripperOpen:initialize
bhs.devilbotz.commands.gripper.GripperOpen:end
bhs.devilbotz.commands.auto.DriveStraightPID:initialize
bhs.devilbotz.commands.auto.DriveStraightPID:end
bhs.devilbotz.commands.assist.PrepareForGroundPickup:Start
bhs.devilbotz.commands.gripper.GripperClose:initialize
bhs.devilbotz.commands.gripper.GripperClose:end
bhs.devilbotz.commands.arm.ArmDown:initialize
bhs.devilbotz.commands.arm.ArmDown:end
bhs.devilbotz.commands.gripper.GripperOpen:initialize
bhs.devilbotz.commands.gripper.GripperOpen:end
bhs.devilbotz.commands.assist.PrepareForGroundPickup:End
bhs.devilbotz.commands.auto.RotateDegrees:initialize
bhs.devilbotz.commands.auto.RotateDegrees:end
bhs.devilbotz.commands.assist.AutoScore:End
```

We can add more debug as needed so that we can understand what is happening in the robot.  Eventually, I'd like to add a way to dynamically enable/disable the debug messages using shuffleboard.

### CODE REVIEW

**Comment by ParkerMeyers** [comment]
> Hi team, since we are experimenting with more complex autonomous/assist routines, I added some debug/trace code so that we can see what is happening in the sequential commands. E.g. this is the output to the console when executing the AutoScore routine:
> 
> ```
> bhs.devilbotz.commands.assist.AutoScore:Start
> bhs.devilbotz.commands.arm.ArmToPosition:initialize
> bhs.devilbotz.commands.arm.ArmToPosition:end
> bhs.devilbotz.commands.gripper.GripperOpen:initialize
> bhs.devilbotz.commands.gripper.GripperOpen:end
> bhs.devilbotz.commands.auto.DriveStraightPID:initialize
> bhs.devilbotz.commands.auto.DriveStraightPID:end
> bhs.devilbotz.commands.assist.PrepareForGroundPickup:Start
> bhs.devilbotz.commands.gripper.GripperClose:initialize
> bhs.devilbotz.commands.gripper.GripperClose:end
> bhs.devilbotz.commands.arm.ArmDown:initialize
> bhs.devilbotz.commands.arm.ArmDown:end
> bhs.devilbotz.commands.gripper.GripperOpen:initialize
> bhs.devilbotz.commands.gripper.GripperOpen:end
> bhs.devilbotz.commands.assist.PrepareForGroundPickup:End
> bhs.devilbotz.commands.auto.RotateDegrees:initialize
> bhs.devilbotz.commands.auto.RotateDegrees:end
> bhs.devilbotz.commands.assist.AutoScore:End
> ```
> 
> We can add more debug as needed so that we can understand what is happening in the robot. Eventually, I'd like to add a way to dynamically enable/disable the debug messages using shuffleboard.

make it so it only does that when the robot is in test mode instead of using a shuffleboard toggle

**Comment by BrownGenius** [comment]
> you could make it so it only does that when the robot is in test mode instead of using a shuffleboard toggle @BrownGenius

In Test mode, how do I select and run the autonomous routines?  Do I need to create a separate autonomous command list for shuffleboard?

**Comment by ParkerMeyers** [comment]
> > you could make it so it only does that when the robot is in test mode instead of using a shuffleboard toggle @BrownGenius
> 
> In Test mode, how do I select and run the autonomous routines? Do I need to create a separate autonomous command list for shuffleboard?

I'm not sure, we'd have to explore it

**Comment by BrownGenius** [comment]
This is an example "DockAndEngage" output:
```
bhs.devilbotz.commands.auto.DockAndEngage:Start
bhs.devilbotz.commands.auto.DriveStraightToDock:initialize
bhs.devilbotz.commands.auto.DriveStraightPID:initialize
bhs.devilbotz.commands.auto.DriveStraightPID:initialize:startAngle: -97.66043090820312 distance: 2.0
bhs.devilbotz.commands.auto.DriveStraightToDock:end
bhs.devilbotz.commands.auto.DriveStraightPID:end:endAngle: -98.35904693603516 distance: 1.9049455093931287
bhs.devilbotz.commands.auto.BalancePID:initialize:startRoll: 0.0
bhs.devilbotz.commands.auto.BalancePID:end:endRoll: 0.0
bhs.devilbotz.commands.auto.RotateDegrees:initialize
bhs.devilbotz.commands.auto.RotateDegrees:initialize:startAngle: -98.98542785644531 --> -8.985427856445312
bhs.devilbotz.commands.auto.RotateDegrees:end:endAngle: -9.902073860168457
bhs.devilbotz.commands.auto.DockAndEngage:End
```

File: src/main/java/bhs/devilbotz/RobotContainer.java [hunk 1]
```diff
@@ -162,7 +155,9 @@ private void configureBindings() {
     SmartDashboard.putData("gripperClose", new GripperClose(gripper));
 
     SmartDashboard.putData(
-        "armScorePiece", new ArmMoveDistance(arm, -10).andThen(new GripperOpen(gripper)));
+        "armScorePiece",
+        new ArmMoveDistance(arm, ArmConstants.POSITION_SCORING_DELTA, gripper)
+            .andThen(new GripperOpen(gripper)));
```
Comment (loafdog) [inline src/main/java/bhs/devilbotz/RobotContainer.java]: i don't think we want gripper to open here.  The button assigned this command does not open gripper.  The dashboard button should do the same thing.  That said, we don't really need this put to dashboard since we put arm commands onto their own shuffleboard tab in buildArmShuffleboardTab()

**Review by loafdog — COMMENTED** [review]
These changes are huge.  And not just adding debug messages.  I don't like that we are changing so much last minute.  We should be changing code as little as possible now.  Focus on letting drivers practice and fix things as we find during driver practice.   I don't agree with merging this PR now.

Test mode is not going to be useful for this.  We want to be able to trace/watch command actions during matches, not just in test mode.  Test mode and live window was intended to be a way for teams to debug individual components on their robot without any of their own code running.  This would let you verify that wiring was good, parts were good.. but it's kind of fallen out of favor for reasons.  Not all vendors don't support it, hard to keep up with new components coming out.  More background here: https://www.chiefdelphi.com/t/have-you-used-the-driver-station-test-mode-successfully-this-year/428685

As an aside, there are already solutions written by FRC community we should look at, two that come to mind are Oblog and Advantage Kit/Scope: 
- https://www.chiefdelphi.com/t/oblog-2023-beta-release/417429
- https://www.chiefdelphi.com/t/introducing-logging-with-advantagekit/398612
- https://www.chiefdelphi.com/search?q=oblog

I played with Oblog a bit last year, never tried Advantage stuff.

File: src/main/java/bhs/devilbotz/RobotContainer.java (hunk 1 above)
Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/RobotContainer.java]: > These changes are huge. And not just adding debug messages. I don't like that we are changing so much last minute. We should be changing code as little as possible now. Focus on letting drivers practice and fix things as we find during driver practice. I don't agree with merging this PR now.

Hi Maciej,

The changes do look "huge" in terms of line change count, but logically the code is much simpler.  The debug code did find a bunch of bugs and corner cases during simulation runs.  The bug fixes are currently interspersed with the debug logic, so I'll split these into two PRs so that it's cleaner.

> Test mode is not going to be useful for this. We want to be able to trace/watch command actions during matches, not just in test mode. Test mode and live window was intended to be a way for teams to debug individual components on their robot without any of their own code running. This would let you verify that wiring was good, parts were good.. but it's kind of fallen out of favor for reasons. Not all vendors don't support it, hard to keep up with new components coming out. More background here: https://www.chiefdelphi.com/t/have-you-used-the-driver-station-test-mode-successfully-this-year/428685
> 
> As an aside, there are already solutions written by FRC community we should look at, two that come to mind are Oblog and Advantage Kit/Scope:
> 
> * https://www.chiefdelphi.com/t/oblog-2023-beta-release/417429
> * https://www.chiefdelphi.com/t/introducing-logging-with-advantagekit/398612
> * https://www.chiefdelphi.com/search?q=oblog
> 
> I played with Oblog a bit last year, never tried Advantage stuff.

I did see the Advantage/Ooblog stuff and it did seem useful, but for tomorrow's testing, I needed something simple.  Nothing beats good ol' console prints for debugging realtime systems.  I'm OK keeping this debug stuff in a branch.  As I mentioned above, I'll push the bug fixes as a separate PR.

Comment (BrownGenius) [reply src/main/java/bhs/devilbotz/RobotContainer.java]: This is what was in mainline.  I just changed it from using the "-10" to the constant and the gripper object for the "auto close" if moving down too close to the frame.  I agree we need to remove a lot of the SmartDashboard calls that have been sprinkled about the code and come up with a more consistent debug methodology.

**Comment by BrownGenius** [comment]
Hi Team, we tested this code on the actual bot and was very useful to trace the various commands.  Until we switch to a more standard debug logging method (e.g. Oblog), we should use this.

**Comment by ParkerMeyers** [comment]
> Hi Team, we tested this code on the actual bot and was very useful to trace the various commands. Until we switch to a more standard debug logging method (e.g. Oblog), we should use this.

I'm not sure you understand what Oblog is, it is a shuffleboard logging utility: https://github.com/Oblarg/Oblog. Anyways, in the long run we should make our own. I've started writing some ideas, we can talk about it in the offseason!

**Review by ParkerMeyers — APPROVED** [review]
Good changes! I like them much better then the smart dashboard. One question, is there a way to disable the debug messages with a toggle in shuffleboard or something?

**Comment by BrownGenius** [comment]
Currently, there isn't a way to enable/disable via shuffleboard, but it should be easy to add that.
